ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/LfValueTranslation.scala

```diff
@@ -0,0 +1,222 @@
+        ),
+    )
+
+  def deserialize[E](raw: Raw.Created[E], verbose: Boolean): CreatedEvent = {
+    val key = LfValueTranslation.Cache.Key(raw.partial.eventId)
+    val create =
+      cache
+        .getIfPresent(key = key)
+        .getOrElse(
+          LfValueTranslation.Cache.Value.Create(
+            argument = ValueSerializer.deserializeValue(raw.createArgument),
+            key = raw.createKeyValue.map(ValueSerializer.deserializeValue)
+          )
+        )
+        .assertCreate()
+    raw.partial.copy(
+      createArguments = Some(
+        toApiRecord(
+          value = create.argument,
+          verbose = verbose,
+          attribute = "create argument",
+        )
+      ),
+      contractKey = create.key.map(
+        key =>
+          toApiValue(
+            value = key,
+            verbose = verbose,
+            attribute = "create key",
+        )
+      ),
+    )
+  }
+
+  def deserialize(raw: Raw.TreeEvent.Exercised, verbose: Boolean): ExercisedEvent = {
+    val key = LfValueTranslation.Cache.Key(raw.partial.eventId)
+    val exercise =
+      cache
+        .getIfPresent(key = key)
+        .getOrElse(
+          LfValueTranslation.Cache.Value.Exercise(
+            argument = ValueSerializer.deserializeValue(raw.exerciseArgument),
+            result = raw.exerciseResult.map(ValueSerializer.deserializeValue)
+          )
+        )
+        .assertExercise()
+    raw.partial.copy(
+      choiceArgument = Some(
+        toApiValue(
+          value = exercise.argument,
+          verbose = verbose,
+          attribute = "exercise argument",
+        )
+      ),
+      exerciseResult = exercise.result.map(
+        result =>
+          toApiValue(
+            value = result,
+            verbose = verbose,
+            attribute = "exercise result",
+        )
+      ),
+    )
+  }
+
+}
+
+object LfValueTranslation {
+
+  type Cache = caching.Cache[Cache.Key, Cache.Value]
+
+  object Cache {
+
+    // To avoid the overhead of measuring the size of objects, we just assign a weight of 1 to values
+    private implicit object `Key Weight` extends caching.Weight[Key] {
+      override def weigh(value: Key): caching.Cache.Size = 0
+    }
+
+    // To avoid the overhead of measuring the size of objects, we just assign a weight of 1 to values
+    private implicit object `Value Weight` extends caching.Weight[Value] {
+      override def weigh(value: Value): caching.Cache.Size = 1
+    }
+
+    def newInstance(configuration: caching.Configuration): Cache =
+      caching.Cache.from(configuration)
+
+    def newInstrumentedInstance(configuration: caching.Configuration, metrics: Metrics): Cache =
+      caching.Cache.from(
+        configuration = configuration,
+        metrics = metrics.daml.index.db.translation.cache,
+      )
+
+    final class UnexpectedTypeException(value: Value)
+        extends RuntimeException(s"Unexpected value $value")
+
+    final case class Key(eventId: String)
```

Comment:
Could this `extends AnyVal`?

Reply:
Also, instead of making `Key` here, perhaps we could make `EventId` a proper type and just use that? (Potentially as a separate piece of work.)

Reply:
> Could this `extends AnyVal`?

Keys must extends `AnyRef`.

Reply:
> Also, instead of making `Key` here, perhaps we could make `EventId` a proper type and just use that? (Potentially as a separate piece of work.)

It is already, I'm using string here because I also have those coming from the database.